Feature: Transaction Insights

alaahd opened the pull request:
### Explanation:

This PR adds the Transactions Insights feature and supporting functionality. This will give users more information about what is happening when sending a transaction to a contract. The core of the functionality is displayed below:

https://user-images.githubusercontent.com/7499938/144235680-024f9aae-586f-4d79-bebc-ad849961d879.mp4

To achieve this, this PR uses https://github.com/trufflesuite/truffle/tree/develop/packages/decoder to decode transaction details which are retrieved from a codefi api

The above screen capture demonstrates the transaction decoding feature and the use of the new modals for viewing and editing account and nickname information.

In addition, the PR adds the transaction decoding to the transaction details modal that is opened when clicking list items in the transaction activity list. This addition also updates the styles/layouts of that modal.

### Manual testing steps:  

This can be done on the Kovan network.

1.   Create a transaction that requires interaction with a contract. Uniswap is a good example
2.   Inspect the "Data" tab to see the transaction details 
3.   Click an address in the transaction details and edit the nickname
4.  Confirm the transaction
5. Click the transaction in the activity list
6. View the transaction information
7. Click "Transaction Data"
8. See the Transaction Data including the nickname you previously edited.

### Screenshots and videos:


https://user-images.githubusercontent.com/7499938/144238081-6a3a2861-25f1-4afa-8ffb-592b1591f087.mp4

Commit:
integration for tx decoding confirmation and history view

alaahd committed:
upgrading @truffle/decoder to latest release 5.1.0

danjm committed:
Update acorn and colors patches

alaahd committed:
feat: remove redundant styling

alaahd committed:
feat: basic integration for nickname components

alaahd committed:
feat: wiring functionality of adding new nickname

alaahd committed:
feat: wire functionality of showing nickname modal

alaahd committed:
feat: link the nickname popover with add/update popover

alaahd committed:
feat: moving forward with address nicknames integration

alaahd committed:
feat: fixing a bug related to passing chainId in addressBook

alaahd committed:
feat: populating memo prop in addressbook entry

alaahd committed:
feat: add explorer link

alaahd committed:
feat: bug fixing update nickname component

alaahd committed:
feat: fix proptypes

alaahd committed:
feat: adding tooltip for copying nickname address

alaahd committed:
featL fix styling for tx-details page

alaahd committed:
feat: optimize code for error handling

alaahd committed:
feat: limiting transaction decoding to tx with data

alaahd committed:
feat: remove tree UI component

alaahd committed:
feat: adding request to check for tx decoding supported networks

alaahd committed:
feat: showing data hex component

alaahd committed:
feat: fix react warnings

alaahd committed:
feat: remove extra margin in tx decoding

alaahd committed:
feat: link tx-decoding addresses with nicknames

danjm commented on ui/components/app/modals/account-nickname-modal/account-nickname-modal.container.js:
```diff
@@ -0,0 +1,30 @@
+import { hideModal, showModal } from '../../../../store/actions';
+import AccountNicknameModal from './account-nickname-modal.component';
+
+function mapStateToProps(state) {
```
We no longer use this pattern for modals. Instead see `metamask-extension/ui/components/app/qr-hardware-popover/qr-hardware-popover.js` as an example: we conditionally render a `<Popover ... />` component, and so the properties can be directly passed from the parent to the popover.

alaahd replied:
Fixed ! we have created a composable functional component responsible for rendering the appropriate nickname modal according to its internal state ! all other modals have been removed.

danjm commented on ui/components/app/modals/update-account-nickname-modal/update-account-nickname-modal.container.js:
```diff
@@ -0,0 +1,36 @@
+import * as actions from '../../../../store/actions';
+import UpdateAccountNicknameModal from './update-account-nickname-modal.component';
+
+function mapStateToProps(state) {
```
Our current pattern is to do all state selection using hooks within the functional component itself. We should do that for `addressBookEntry` as well

alaahd replied:
Fixed.

danjm commented on ui/components/app/transaction-decoding/components/decoding/address/address.container.js:
```diff
@@ -0,0 +1,34 @@
+import * as actions from '../../../../../../store/actions';
+import Address from './address.component';
+
+function mapStateToProps(state, ownProps) {
```
Let's move this into the functional component, and get `recipientNickname` by selecting state using react-redux hooks

alaahd replied:
Fixed.

danjm commented on ui/components/app/transaction-decoding/components/ui/copy-raw-data/index.scss:
```diff
@@ -0,0 +1,27 @@
+        background-color: transparent;
+
+        &:hover {
+        background-color: #f2f3f4;
```
We use sass variables for colours. See `ui/css/design-system/colors.scss`

gnidan replied:
`#f2f3f4` has a variable (`$Gray-000`), but we notice `#ededed` has no variable and is used elsewhere. What should we do? Make a new variable? Or find an existing variable that is similar?

danjm replied:
we should find an existing variable which is similar, can be done as part of the PR to address https://github.com/MetaMask/metamask-extension/pull/12881#issuecomment-983474180

danjm commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,249 @@
+    (async () => {
+      setLoading(true);
+      try {
+        const networks = await fetchWithCache(
```
What is this fails due to network or server failure? Do the error messages below cover it?

alaahd replied:
It looks like fetch-with-cache handles error appropriately ! so we have reverted all redundant error throwing !

danjm commented on ui/components/app/transaction-list-item-details/transaction-list-item-details.component.js:
```diff
@@ -4,6 +4,7 @@ import copyToClipboard from 'copy-to-clipboard';
 import { getBlockExplorerLink } from '@metamask/etherscan-link';
```
I think it might be best to move the transaction list related changes to a separate PR

alaahd replied:
I will be happy to do that, how do you want to deal with that considering the timing constraints ?

danjm replied:
We decided not to do this

danjm commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,249 @@
+import React, { useContext, useEffect, useState, useRef } from 'react';
```
Was the original plan to have this component as a separate module that is imported?

alaahd replied:
Ultimately this was our initial plan ! but considering the timing we decided to extract later.

danjm commented on ui/components/app/transaction-decoding/components/ui/copy-raw-data/copy-raw-data.component.js:
```diff
@@ -0,0 +1,34 @@
+          <div className="copy-raw-data__icon">
+            <CopyIcon size={12} color="#BBC0C5" />
+          </div>
+          <div className="copy-raw-data__label">Copy raw transaction data</div>
```
`Copy raw transaction data` should be translated

danjm commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,249 @@
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    (async () => {
```
For example, see this swaps util: `ui/pages/swaps/swaps.util.js`

It contains all the logic for fetching that is used throughout the swaps feature

danjm replied:
could be put in `ui/components/app/transaction-decoding/transaction-decoding.util.js`

danjm reviewed (changes requested):
So there are few changes we need to this PR.

Here is how I would handle them, in order of priority:

1. Rebase onto develop, so that the changes from this PR are not duplicated https://github.com/MetaMask/metamask-extension/pull/12632
2. Refactor to use the current "Popover" pattern for showing modals, and then refactor to use react-redux hooks within functional components instead of "container" components. The "Popover" pattern is to import the respective popover component into the appropriate parent, pass the props from that parent directly to the component, and then conditionally render the component (and not to use redux for storing the data or indicating whether the popover should be shown). `ui/components/app/qr-hardware-popover/qr-hardware-popover.js` is an example
3. It is probably best to have the confirm screen changes and the transaction details changes in separate PRs. This PR is big. Separate PRs will help code reviewers focus and it is possible that some of the code review can be parallelized if in separate PRs
4. Move the logic in the useEffect of the transaction-decoding component to help methods in a util file
5. Address the small comments

There is also some clean up that needs to happen of css, but that can be separate.

Commit:
Ensure messages get translated

alaahd committed:
Merge pull request #1 from gnidan/feature/address-nicknames

Ensure messages get translated

alaahd committed:
fix: fixing issue with transaltion

alaahd committed:
fix: revert fetchWithCache modifications

alaahd commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,249 @@
+        const networks = await fetchWithCache(
+          FETCH_SUPPORTED_NETWORKS_URI,
+          { method: 'GET' },
+          { suppressAutomaticThrowing: true },
```
Turns out it wasn't necessary anyway ! we have reverted the changes for fetch-with-cache.js

gnidan commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,249 @@
+        );
+
+        if (!networks) {
+          throw new Error(
```
I believe I got the relevant ones (seems like it's common to leave things like `Error: ...` untranslated elsewhere).

alaahd commented on ui/components/ui/sender-to-recipient/sender-to-recipient.container.js:
```diff
@@ -0,0 +1,18 @@
+import * as actions from '../../../store/actions';
+import SenderToRecipient from './sender-to-recipient.component';
+
+function mapDispatchToProps(dispatch) {
```
Can you provide me with an example on the new approach where I can follow ?

gnidan replied:
(we found the example you linked after we asked, sorry!)

danjm replied:
this has been properly addressed

alaahd commented on ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:
```diff
@@ -120,7 +120,10 @@ const mapStateToProps = (state, ownProps) => {
     shortenAddress(toChecksumHexAddress(toAddress));
 
   const checksummedAddress = toChecksumHexAddress(toAddress);
-  const addressBookObject = addressBook[checksummedAddress];
```
Yes! I fixed a bug :)

alaahd replied:
The current implementation is not attributing to chainId when returning the addressBook object.

gnidan commented:
I have read the CLA Document and I hereby sign the CLA

alaahd commented on ui/components/ui/nickname-popover/nickname-popover.component.js:
```diff
@@ -0,0 +1,96 @@
+import React, { useCallback, useContext } from 'react';
```
Done !

alaahd committed:
fix: sass color variable

danjm committed:
Moving to popover pattern

danjm committed:
Remove changes from modal.js

gnidan commented:
Gabe raised the issue that the "copy transaction button" copies the text `"[object Object]"`

alaahd committed:
fix: refactor nickname popovers

alaahd committed:
Merge branch 'feature/address-nicknames' of github.com:alaahd/metamask-extension into feature/address-nicknames

Commit:
Remove unused package @truffle/source-map-utils

gnidan commented on ui/components/app/transaction-decoding/components/decoding/address/address.component.js:
```diff
@@ -0,0 +1,122 @@
+    null,
+  );
+
+  let tooltipHtml = <p>{t('copiedExclamation')}</p>;
+  if (!addressCopied) {
+    tooltipHtml = '';
+  }
```
This `tooltipHtml` doesn't get used for anything - should it?

alaahd replied:
you are right, I just removed it !

Commit:
Omit value for boolean attributes

Commit:
Fix props reading in CopyRawData

Commit:
Fix lint errors in TransactionDecoding

- Remove unused import
- Reorder imports
- Address conflict between caught `error` and error state flag by
  renaming state flag to `hasError`
- Fix requestUrl identifier casing and use of template string
- Ensure `useEffect` gets passed the deps it needs
- Add scope braces around case statement where it's needed
- Omit literal `true` for boolean jsx attribute
- Refactor nested ternary as `if` statements

Commit:
Fix linting for TransactionListItemDetails

- Remove unused import
- Fix import spacing
- Remove unused prop dereference
- Fix string interpolation for translated From/To

Commit:
Omit value for boolean attribute

Commit:
Ensure const gets declared before it's used

Commit:
Fix linting for ConfirmTransactionBase

- Remove unused prop chainId
- Stop destructuring an unused field

alaahd committed:
fix: refactor usage of nicknames popovers in send-content-container

alaahd committed:
Merge branch 'feature/address-nicknames' of github.com:alaahd/metamask-extension into feature/address-nicknames

alaahd committed:
fix: remove extra prop updateAccountNicknameModal

alaahd committed:
fix: refactor code for address.component

Commit:
Ensure NicknamePopovers always returns component

alaahd commented:
I have read the CLA Document and I hereby sign the CLA

alaahd committed:
fix: remove extra tooltip

Commit:
Fix linting for NicknamePopover component

- Fix useCallback deps
- Switch ternary to logical-or

Commit:
Fix linting for SenderToRecipient

... by fixing import order

Commit:
Merge remote-tracking branch 'alaahd/feature/address-nicknames' into feature/address-nicknames

Commit:
Remove unused addressCopied state

danjm committed:
Delete empty file

alaahd committed:
fix: remove sender-to-recipient.container

alaahd committed:
Merge branch 'feature/address-nicknames' of github.com:alaahd/metamask-extension into feature/address-nicknames

alaahd committed:
fix: refactor usage of nickname popovers in confirm-page-container

alaahd committed:
fix: bug related to state variable

danjm committed:
Stylelint fix

danjm committed:
Lint fix

Commit:
Change "Total Amount" to "Total"

Commit:
Merge remote-tracking branch 'alaahd/feature/address-nicknames' into feature/address-nicknames

danjm committed:
Lint fix locales

danjm committed:
Update address-book.spec.js

danjm committed:
e2e test update

danjm committed:
Update e2e tests

gnidan commented:
The robot failed to make this announcement:
<blockquote>
<details><summary>Builds ready [17bcd58]</summary><ul><li>builds: <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds/metamask-chrome-10.6.4.zip">chrome</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds/metamask-firefox-10.6.4.zip">firefox</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds/metamask-opera-10.6.4.zip">opera</a></li>
<li>builds (beta): <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-beta/metamask-beta-chrome-10.6.4.zip">chrome</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-beta/metamask-beta-firefox-10.6.4.zip">firefox</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-beta/metamask-beta-opera-10.6.4.zip">opera</a></li>
<li>builds (flask): <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-flask/metamask-flask-chrome-10.6.4.zip">chrome</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-flask/metamask-flask-firefox-10.6.4.zip">firefox</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/builds-flask/metamask-flask-opera-10.6.4.zip">opera</a></li>
<li>build viz: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/build-viz/index.html">Build System</a></li>
<li>code coverage: <a href="https://405227-42009758-gh.circle-artifacts.com/0/coverage/index.html">Report</a></li>
<li>storybook: <a href="https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html">Storybook</a></li>
<li><a href="https://circleci.com/gh/MetaMask/metamask-extension/405227#artifacts/containers/0">all artifacts</a></li>
<li><details>
       <summary>bundle viz:</summary>
       <ul><li>background: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-0.html">0</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-1.html">1</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-2.html">2</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-3.html">3</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-4.html">4</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-5.html">5</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-6.html">6</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/background-7.html">7</a></li><li>common: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/common-0.html">0</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/common-1.html">1</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/common-2.html">2</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/common-3.html">3</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/common-4.html">4</a></li><li>content-script: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/content-script-0.html">0</a></li><li>contentscript: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/contentscript.html">contentscript</a></li><li>disable-console: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/disable-console.html">disable-console</a></li><li>inpage: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/inpage.html">inpage</a></li><li>phishing-detect: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/phishing-detect.html">phishing-detect</a></li><li>policy-load: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/policy-load.html">policy-load</a></li><li>sentry-install: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/sentry-install.html">sentry-install</a></li><li>ui: <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-0.html">0</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-1.html">1</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-10.html">10</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-2.html">2</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-3.html">3</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-4.html">4</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-5.html">5</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-6.html">6</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-7.html">7</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-8.html">8</a>, <a href="https://405227-42009758-gh.circle-artifacts.com/0/build-artifacts/source-map-explorer/ui-9.html">9</a></li></ul>
     </details></li></ul></details>

<details><summary>Page Load Metrics (1057 ± 47 ms)</summary><table><thead><tr><th>Platform</th><th>Page</th><th>Metric</th><th>Min (ms)</th><th>Max (ms)</th><th>Average (ms)</th><th>StandardDeviation (ms)</th><th>MarginOfError (ms)</th></tr></thead><tbody><tr><td rowspan="4">Chrome</td><td rowspan="4">Home</td><td>firstPaint</td><td align="right">55</td><td align="right">1002</td><td align="right">177</td><td align="right">227</td><td align="right">109</td></tr><tr><td>domContentLoaded</td><td align="right">972</td><td align="right">1433</td><td align="right">1057</td><td align="right">98</td><td align="right">47</td></tr><tr><td>load</td><td align="right">972</td><td align="right">1433</td><td align="right">1057</td><td align="right">98</td><td align="right">47</td></tr><tr><td>domInteractive</td><td align="right">972</td><td align="right">1433</td><td align="right">1057</td><td align="right">98</td><td align="right">47</td></tr></tbody></table></details>

### highlights:
<details>
    <summary>storybook</summary>
    
- [ui/components/app/metamask-translation/metamask-translation.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-components-app-metamask-translation-metamask-translation-stories-js)
- [ui/components/app/transaction-list/transaction-list.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-components-app-transaction-list-transaction-list-stories-js)
- [ui/components/ui/nickname-popover/nickname-popover.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-components-ui-nickname-popover-nickname-popover-stories-js)
- [ui/components/ui/update-nickname-popover/update-nickname-popover.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-components-ui-update-nickname-popover-update-nickname-popover-stories-js)
- [ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-add-suggested-token-confirm-add-suggested-token-stories-js)
- [ui/pages/confirm-approve/confirm-approve.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-approve-confirm-approve-stories-js)
- [ui/pages/confirm-deploy-contract/confirm-deploy-contract.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-deploy-contract-confirm-deploy-contract-stories-js)
- [ui/pages/confirm-encryption-public-key/confirm-encryption-public-key.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-encryption-public-key-confirm-encryption-public-key-stories-js)
- [ui/pages/confirm-import-token/confirm-import-token.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-import-token-confirm-import-token-stories-js)
- [ui/pages/confirm-send-ether/confirm-send-ether.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-send-ether-confirm-send-ether-stories-js)
- [ui/pages/confirm-send-token/confirm-send-token.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-send-token-confirm-send-token-stories-js)
- [ui/pages/confirm-token-transaction-base/confirm-token-transaction-base.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-token-transaction-base-confirm-token-transaction-base-stories-js)
- [ui/pages/confirm-transaction-base/confirm-transaction-base.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-confirm-transaction-base-confirm-transaction-base-stories-js)
- [ui/pages/send/send-content/send-content.stories.js](https://405227-42009758-gh.circle-artifacts.com/0/storybook/index.html?path=/story/ui-pages-send-send-content-send-content-stories-js)
  </details>
</blockquote>

danjm commented:
https://405327-42009758-gh.circle-artifacts.com/0/builds/metamask-brave-10.6.4.zip

https://405327-42009758-gh.circle-artifacts.com/0/builds/metamask-chrome-10.6.4.zip

https://405327-42009758-gh.circle-artifacts.com/0/builds/metamask-firefox-10.6.4.zip

danjm commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -632,25 +637,52 @@ export default class ConfirmTransactionBase extends Component {
             <span className="confirm-page-container-content__function-type">
-              {functionType}
+              {`${functionType} (${
+                params ? params.map(({ type }) => type).join(', ') : null
```
`null` will be rendered as a string here, this should be an empty string instead

danjm commented:
Here are a number of issues we need to resolve, but I think it is okay if they are resolved in a separate PR (as long as it is prioritized immediately after we merge this and before we add any other features):

**CSS**
- should be resolvable without having to specify `.disclosure` as a child of `_data`, could require the addition of a different class. https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-37b35ddf3d088e4a3707b8fbc05c06ae8aeba47b1ac77905359865d3760231a8R24
- nit: https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-c85f241de885a39d520383b6f4b7b713b58d7bc333201e3f39584e31f0dc46a4R20 this is conventionally done like this:
```
className={classnames(
  'transaction-breakdown-row',
  className,
  {
    'transaction-breakdown-row--with-bottom-border': divider
  }
)}
```
- https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-c4c2d6f4e4c3c02fd8d87af59450f1416887397f3f6fe2ac72c7fda5c76764a4R2 should follow the BEM pattern
- colors need to be variables: https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-a975a059e7e6451dc1f8e13e426f15113bc9d7fedad48dcb332c5cf6cd6300f0R19
- we should instead use BEM patterns, more classes and conditional classes, instead of extension use f html tag selectors and child selectors in https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-3edbf77b1c98a7db9d30cb23550dea1b466e3009d919e261d60ba1bbd7050448R21-R164
- ideally our css shouldn't reference classes of other components, instead the other component should be configurable and manage its own css under that configuration https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-503fa60e87628a2c02738c80d14b3fe666a3bc8574eed53af6a41a97f8cf9e62R6
- add another class, using BEM style, instead of child selectors: https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-f7d836e9f6f1a3588f842b9ee5a55e3f3d282672d2ddf8c3e7cc70eac8fceeccR6

**JavaScript**
- move fetchWithCache calls to a helper function https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-0d33bff56e4fdad212852eb6845d22ea07ca0ab70437ff224e8489df4e92d35eR37-R58

**Tests:**
- add unit test to replace this one that was removed https://github.com/MetaMask/metamask-extension/pull/12881/files#diff-d4ab175dc9f8c76a2bafdfc82421ebe30922f7dddd3b061150ab16e8c0c449a3
- add some unit tests for the new components
- add some unit tests for the transaction decoding util

danjm committed:
Fix issue where absence of function params in data hex tab would result in rendering a  string

danjm committed:
Fix border radius, and width and height in small notification windows, of the update-nickname-popover

danjm commented on ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:
```diff
@@ -15,6 +15,7 @@ export default class ConfirmPageContainerContent extends Component {
   static propTypes = {
```
The changes in this file allow for the addition of the "Hex" tab alongside the "Data" tab as shown here:
![Screenshot from 2021-12-01 09-39-08](https://user-images.githubusercontent.com/7499938/144240129-f3c17507-93a5-4926-8936-adf2f028a250.png)

danjm commented on ui/components/app/modals/nickname-popovers/nickname-popovers.component.js:
```diff
@@ -0,0 +1,72 @@
+import React, { useState } from 'react';
```
This file controls the display of the two nickname popovers:

![Screenshot from 2021-12-01 09-40-45](https://user-images.githubusercontent.com/7499938/144240300-e198ca51-926c-4e23-9641-33ff5711b890.png)
![Screenshot from 2021-12-01 09-40-41](https://user-images.githubusercontent.com/7499938/144240301-aa85ed3e-56d0-4d41-a07a-1537bbde03db.png)

danjm commented on ui/components/app/transaction-decoding/components/decoding/address/address.component.js:
```diff
@@ -0,0 +1,72 @@
+import React, { useState } from 'react';
```
This component controls the display and behaviour of each icon+address in the decoded transaction information:
![Screenshot from 2021-12-01 09-41-43](https://user-images.githubusercontent.com/7499938/144240456-ab7d1b34-0f41-4550-9838-4af261273b1f.png)

danjm commented on ui/components/app/transaction-decoding/components/ui/copy-raw-data/copy-raw-data.component.js:
```diff
@@ -0,0 +1,38 @@
+import React, { useContext } from 'react';
```
This is for the "Copy raw transaction data" component in the hex tab:

![Screenshot from 2021-12-01 09-42-44](https://user-images.githubusercontent.com/7499938/144240612-020434c8-3a4f-42b8-8931-118de59272d8.png)

danjm commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,224 @@
+import React, { useContext, useEffect, useState } from 'react';
```
This component is responsible for decoded transaction data, the primarily highlight of this feature/PR:

![Screenshot from 2021-12-01 09-43-33](https://user-images.githubusercontent.com/7499938/144240775-1f436119-66a6-4715-9954-797da09c0b00.png)

danjm commented on ui/components/app/transaction-list-item-details/transaction-list-item-details.component.js:
```diff
@@ -4,16 +4,17 @@ import copyToClipboard from 'copy-to-clipboard';
 import { getBlockExplorerLink } from '@metamask/etherscan-link';
```
The changes in this file are responsible for the layout changes in the transaction details modal:

![Screenshot from 2021-12-01 09-45-09](https://user-images.githubusercontent.com/7499938/144241045-6ce9e14f-9608-4eaf-9a4a-fc75b76f7051.png)

danjm commented on ui/components/ui/disclosure/disclosure.js:
```diff
@@ -0,0 +1,51 @@
+import React, { useState, useRef, useEffect } from 'react';
```
This component handles the togglable dropdown ux and functionality shown below:

https://user-images.githubusercontent.com/7499938/144241214-e236f048-7206-4662-98e4-713873c6685a.mp4

danjm commented on ui/components/ui/sender-to-recipient/sender-to-recipient.component.js:
```diff
@@ -8,6 +8,7 @@ import { shortenAddress } from '../../../helpers/utils/util';
 import AccountMismatchWarning from '../account-mismatch-warning/account-mismatch-warning.component';
```
The changes in this file account for the new behaviour of clicking the address shown at the top of the confirm screen (and in the transaction details modal):


https://user-images.githubusercontent.com/7499938/144241598-f4b01da6-6820-494d-ac67-8b9e0a04ecb1.mp4

segun commented on ui/components/app/transaction-decoding/transaction-decoding.component.js:
```diff
@@ -0,0 +1,224 @@
+        });
+
+        // fake await
+        await new Promise((resolve) => {
```
Why is this 500ms pause needed?

danjm replied:
that's a good question

alaahd replied:
we need to remove it, we had it temporarily as sometimes we get response faster than expected where we can not show the loading spinner ! but we can totally remove it.

danjm replied:
removed with 5ffefd449

jpuri commented on ui/components/ui/nickname-popover/nickname-popover.component.js:
```diff
@@ -27,31 +33,51 @@ const NicknamePopover = ({ address, onClose = null, onAdd = null }) => {
+            }}
+            target="_blank"
+            rel="noopener noreferrer"
+            title={t('etherscanView')}
```
This look un-necessary
target="_blank"
rel="noopener noreferrer"

alaahd replied:
yes I agree, but the same pattern is being used in the codebase for external links in some places, a similar pattern can be found in `/ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js` line 417. happy to remove it if it has no purpose.

danjm replied:
So these are actually needed.

![Screenshot from 2021-12-01 11-27-32](https://user-images.githubusercontent.com/7499938/144257598-166ddb57-de87-45ed-b554-b6d3d1876942.png)

danjm replied:
Good point @jpuri
I created a ticket for this: https://github.com/MetaMask/metamask-extension/issues/12923

jpuri commented on app/images/icons/collapse.svg:
```diff
@@ -0,0 +1,26 @@
+<rect width="413" height="679" fill="white" transform="translate(-57 -402)"/>
+</clipPath>
+</defs>
+</svg>
```
```suggestion
<svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#E5E5E5" d="M0 0h16v16H0z"/><g clip-path="url(#a)"><path fill="#fff" d="M-57-402h413v679H-57z"/><g filter="url(#b)"><path d="M-31-376a8 8 0 0 1 8-8h344a8 8 0 0 1 8 8v628a8 8 0 0 1-8 8H-23a8 8 0 0 1-8-8v-628Z" fill="#fff"/></g><circle cx="8" cy="8" r="7.5" fill="#F2F3F4" stroke="#D6D9DC"/><path stroke="#D6D9DC" stroke-linecap="round" d="M4.1 7.833h7.667"/></g><path stroke="#000" d="M-2918.5-1621.5h9747v4874h-9747z"/><defs><clipPath id="a"><path fill="#fff" transform="translate(-57 -402)" d="M0 0h413v679H0z"/></clipPath><filter id="b" x="-58" y="-409" width="414" height="698" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB"><feFlood flood-opacity="0" result="BackgroundImageFix"/><feColorMatrix in="SourceAlpha" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/><feOffset dy="2"/><feGaussianBlur stdDeviation="13.5"/><feColorMatrix values="0 0 0 0 0.12594 0 0 0 0 0.182502 0 0 0 0 0.305378 0 0 0 0.26 0"/><feBlend in2="BackgroundImageFix" result="effect1_dropShadow_0_1"/><feBlend in="SourceGraphic" in2="effect1_dropShadow_0_1" result="shape"/></filter></defs></svg>```

jpuri commented on app/images/icons/collapse.svg:
```diff
@@ -0,0 +1,26 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+<rect width="16" height="16" fill="#E5E5E5"/>
+<g clip-path="url(#clip0_0_1)">
+<rect width="413" height="679" transform="translate(-57 -402)" fill="white"/>
+<g filter="url(#filter0_d_0_1)">
+<path d="M-31 -376C-31 -380.418 -27.4183 -384 -23 -384H321C325.418 -384 329 -380.418 329 -376V252C329 256.418 325.418 260 321 260H-23C-27.4183 260 -31 256.418 -31 252V-376Z" fill="white"/>
+</g>
+<circle cx="8" cy="8" r="7.5" fill="#F2F3F4" stroke="#D6D9DC"/>
+<line x1="4.1001" y1="7.83325" x2="11.7668" y2="7.83325" stroke="#D6D9DC" stroke-linecap="round"/>
+</g>
+<rect x="-2918.5" y="-1621.5" width="9747" height="4874" stroke="black"/>
+<defs>
+<filter id="filter0_d_0_1" x="-58" y="-409" width="414" height="698" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
+<feFlood flood-opacity="0" result="BackgroundImageFix"/>
+<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
+<feOffset dy="2"/>
+<feGaussianBlur stdDeviation="13.5"/>
+<feColorMatrix type="matrix" values="0 0 0 0 0.12594 0 0 0 0 0.182502 0 0 0 0 0.305378 0 0 0 0.26 0"/>
+<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_0_1"/>
+<feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_0_1" result="shape"/>
+</filter>
+<clipPath id="clip0_0_1">
+<rect width="413" height="679" fill="white" transform="translate(-57 -402)"/>
+</clipPath>
+</defs>
+</svg>
```
Minified version:

```suggestion
<svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#E5E5E5" d="M0 0h16v16H0z"/><g clip-path="url(#a)"><path fill="#fff" d="M-57-402h413v679H-57z"/><g filter="url(#b)"><path d="M-31-376a8 8 0 0 1 8-8h344a8 8 0 0 1 8 8v628a8 8 0 0 1-8 8H-23a8 8 0 0 1-8-8v-628Z" fill="#fff"/></g><circle cx="8" cy="8" r="7.5" fill="#F2F3F4" stroke="#D6D9DC"/><path stroke="#D6D9DC" stroke-linecap="round" d="M4.1 7.833h7.667"/></g><path stroke="#000" d="M-2918.5-1621.5h9747v4874h-9747z"/><defs><clipPath id="a"><path fill="#fff" transform="translate(-57 -402)" d="M0 0h413v679H0z"/></clipPath><filter id="b" x="-58" y="-409" width="414" height="698" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB"><feFlood flood-opacity="0" result="BackgroundImageFix"/><feColorMatrix in="SourceAlpha" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/><feOffset dy="2"/><feGaussianBlur stdDeviation="13.5"/><feColorMatrix values="0 0 0 0 0.12594 0 0 0 0 0.182502 0 0 0 0 0.305378 0 0 0 0.26 0"/><feBlend in2="BackgroundImageFix" result="effect1_dropShadow_0_1"/><feBlend in="SourceGraphic" in2="effect1_dropShadow_0_1" result="shape"/></filter></defs></svg>
```

danjm replied:
Actually, this will get minified by our build scripts. So might be best to leave it as is for readability.

jpuri commented on app/images/icons/expand.svg:
```diff
@@ -0,0 +1,28 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+<rect width="16" height="16" fill="#E5E5E5"/>
+<g clip-path="url(#clip0_0_1)">
+<rect width="413" height="679" transform="translate(-57 -546)" fill="white"/>
+<g filter="url(#filter0_d_0_1)">
+<path d="M-31 -520C-31 -524.418 -27.4183 -528 -23 -528H321C325.418 -528 329 -524.418 329 -520V108C329 112.418 325.418 116 321 116H-23C-27.4183 116 -31 112.418 -31 108V-520Z" fill="white"/>
+</g>
+<line x1="7.5" y1="-128" x2="7.49999" y2="8" stroke="#D6D9DC"/>
+<circle cx="8" cy="8" r="7.5" fill="white" stroke="#D6D9DC"/>
+<line x1="4.1001" y1="7.83325" x2="11.7668" y2="7.83325" stroke="#D6D9DC" stroke-linecap="round"/>
+<line x1="8.1001" y1="4.16699" x2="8.1001" y2="11.8337" stroke="#D6D9DC" stroke-linecap="round"/>
+</g>
+<rect x="-2918.5" y="-1765.5" width="9747" height="4874" stroke="black"/>
+<defs>
+<filter id="filter0_d_0_1" x="-58" y="-553" width="414" height="698" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
+<feFlood flood-opacity="0" result="BackgroundImageFix"/>
+<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
+<feOffset dy="2"/>
+<feGaussianBlur stdDeviation="13.5"/>
+<feColorMatrix type="matrix" values="0 0 0 0 0.12594 0 0 0 0 0.182502 0 0 0 0 0.305378 0 0 0 0.26 0"/>
+<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_0_1"/>
+<feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_0_1" result="shape"/>
+</filter>
+<clipPath id="clip0_0_1">
+<rect width="413" height="679" fill="white" transform="translate(-57 -546)"/>
+</clipPath>
+</defs>
+</svg>
```
```suggestion
<svg width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill="#E5E5E5" d="M0 0h16v16H0z"/><g clip-path="url(#a)"><path fill="#fff" d="M-57-546h413v679H-57z"/><g filter="url(#b)"><path d="M-31-520a8 8 0 0 1 8-8h344a8 8 0 0 1 8 8v628a8 8 0 0 1-8 8H-23a8 8 0 0 1-8-8v-628Z" fill="#fff"/></g><path stroke="#D6D9DC" d="M7.5-128V8"/><circle cx="8" cy="8" r="7.5" fill="#fff" stroke="#D6D9DC"/><path stroke="#D6D9DC" stroke-linecap="round" d="M4.1 7.833h7.667M8.1 4.167v7.667"/></g><path stroke="#000" d="M-2918.5-1765.5h9747v4874h-9747z"/><defs><clipPath id="a"><path fill="#fff" transform="translate(-57 -546)" d="M0 0h413v679H0z"/></clipPath><filter id="b" x="-58" y="-553" width="414" height="698" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB"><feFlood flood-opacity="0" result="BackgroundImageFix"/><feColorMatrix in="SourceAlpha" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/><feOffset dy="2"/><feGaussianBlur stdDeviation="13.5"/><feColorMatrix values="0 0 0 0 0.12594 0 0 0 0 0.182502 0 0 0 0 0.305378 0 0 0 0.26 0"/><feBlend in2="BackgroundImageFix" result="effect1_dropShadow_0_1"/><feBlend in="SourceGraphic" in2="effect1_dropShadow_0_1" result="shape"/></filter></defs></svg>
```

danjm replied:
Actually, this will get minified by our build scripts. So might be best to leave it as is for readability.

danjm committed:
Remove fake await

jpuri reviewed (dismissed):
👍 nice work

Please address feedbacks before merging.

darkwing commented on ui/components/app/transaction-decoding/components/decoding/address/address.component.js:
```diff
@@ -0,0 +1,72 @@
+        className="address__name"
+        onClick={() => setShowNicknamePopovers(true)}
+      >
+        {addressOnly
```
Not a blocker but I'd prefer if this was assigned as a variable above the JSX.

danjm replied:
done in 671e006b1

darkwing commented on ui/components/app/transaction-decoding/components/ui/copy-raw-data/copy-raw-data.component.js:
```diff
@@ -0,0 +1,38 @@
+
+  return (
+    <div className="copy-raw-data">
+      <Tooltip position="right" title={copied ? 'Copied!' : ''}>
```
We should use `copiedExclamation` localization here

danjm replied:
1efaa13e6

darkwing commented on ui/components/app/transaction-decoding/components/ui/copy-raw-data/index.scss:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  &__icon {
+    padding-right: 6px;
```
`padding-inline-end` would be more friend to RTL.

danjm replied:
1efaa13e6

darkwing commented on ui/components/app/transaction-decoding/index.scss:
```diff
@@ -0,0 +1,169 @@
+      font-size: 14px;
+      position: relative;
+      margin: 12px 0 4px;
+      padding-left: 22px;
```
`padding-inline-start` would be good here.

danjm replied:
1efaa13e6

darkwing commented on ui/components/app/transaction-decoding/index.scss:
```diff
@@ -0,0 +1,169 @@
+      cursor: pointer;
+
+      & + .tx-insight-content {
+        padding-left: 14px;
```
`padding-inline-start` would be good here.

danjm replied:
1efaa13e6

danjm committed:
Clean up

danjm committed:
Clean up
